pkg/cluster/storage.go

```diff
@@ -56,18 +73,85 @@ func (c *Cluster) storageBootstrap(ctx context.Context) error {
 		return err
 	}
 
-	value, err := storageClient.Get(ctx, storageKey(c.config.Token))
-	if err == client.ErrNotFound {
-		c.saveBootstrap = true
-		return nil
-	} else if err != nil {
-		return err
+	token := c.config.Token
+	if token == "" {
+		tokenFromFile, err := getTokenFromFile(c.runtime.ServerToken, c.config.DataDir)
+		if err != nil {
+			return err
+		}
+		token = tokenFromFile
 	}
 
-	data, err := decrypt(c.config.Token, value.Data)
+	value, emptyKey, err := c.getBootstrapKeyFromStorage(ctx, storageClient, token)
+	if err != nil {
+		return err
+	}
+	if value == nil {
+		return nil
+	}
+	if emptyKey {
+		token = ""
+	}
+	data, err := decrypt(token, value.Data)
 	if err != nil {
 		return err
 	}
 
 	return bootstrap.Read(bytes.NewBuffer(data), &c.runtime.ControlRuntimeBootstrap)
 }
+
+// getBootstrapKeyFromStorage will list all keys that has prefix /bootstrap and will check for key that is
+// hashed with empty string and will check for any key that is hashed by different token than the one
+// passed to it, it will return error if it finds a key that is hashed with different token and will return
+// value if it finds the key hashed by passed token or empty string
+func (c *Cluster) getBootstrapKeyFromStorage(ctx context.Context, storageClient client.Client, token string) (*client.Value, bool, error) {
+	emptyStringKey := storageKey("")
+	tokenKey := storageKey(token)
+
+	oldBootstrapValues, err := storageClient.List(ctx, "/bootstrap", 0)
+	if err != nil {
+		if err != client.ErrNotFound {
+			c.saveBootstrap = true
+			return nil, false, nil
+		}
+		return nil, false, err
+	}
+	for _, v := range oldBootstrapValues {
```

Comment:
Would like to keep the process deterministic and avoid looping over all values. Should probably error if `/bootstrap` contains more than one value, and explicitly migrate an `emptyStringKey` if it exists.